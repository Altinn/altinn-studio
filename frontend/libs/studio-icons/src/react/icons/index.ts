export { AccordionIcon } from './AccordionIcon';
export { ArrayIcon } from './ArrayIcon';
export { BooleanIcon } from './BooleanIcon';
export { CheckboxIcon } from './CheckboxIcon';
export { CombinationIcon } from './CombinationIcon';
export { ConfirmationTaskIcon } from './ConfirmationTaskIcon';
export { DataTaskIcon } from './DataTaskIcon';
export { DefinitionIcon } from './DefinitionIcon';
export { ElementIcon } from './ElementIcon';
export { EndEventIcon } from './EndEventIcon';
export { FeedbackTaskIcon } from './FeedbackTaskIcon';
export { GiteaIcon } from './GiteaIcon';
export { GitHubIcon } from './GitHubIcon';
export { GroupIcon } from './GroupIcon';
export { LikertIcon } from './LikertIcon';
export { LongTextIcon } from './LongTextIcon';
export { NavBarIcon } from './NavBarIcon';
export { NumberIcon } from './NumberIcon';
export { ObjectIcon } from './ObjectIcon';
export { TextIcon } from './TextIcon';
export { PaymentDetailsIcon } from './PaymentDetailsIcon';
export { PaymentTaskIcon } from './PaymentTaskIcon';
export { PropertyIcon } from './PropertyIcon';
export { RadioButtonIcon } from './RadioButtonIcon';
export { ReferenceIcon } from './ReferenceIcon';
export { RepeatingGroupIcon } from './RepeatingGroupIcon';
export { SectionHeaderWarningIcon } from './SectionHeaderWarningIcon';
export { SelectIcon } from './SelectIcon';
export { ShortTextIcon } from './ShortTextIcon';
export { SignTaskIcon } from './SignTaskIcon';
export { SlackIcon } from './SlackIcon';
export { StringIcon } from './StringIcon';
export { TitleIcon } from './TitleIcon';
