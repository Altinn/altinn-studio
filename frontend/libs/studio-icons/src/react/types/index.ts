export type { IconProps } from './IconProps';
