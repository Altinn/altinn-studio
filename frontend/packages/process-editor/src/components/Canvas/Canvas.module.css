:root {
  --mask-icon-size: 0 0 /46px 46px;
}

.wrapper {
  display: flex;
  flex: 1;
  margin-top: 18px;
}

div[class*='bpmn-icon-task-generic'] {
  background-color: var(--palette-entry-color);
}

div[class*='bpmn-icon-task-generic']:hover {
  background-color: var(--palette-entry-hover-color);
}

div[class*='bpmn-icon-data-task'] {
  -webkit-mask: url('../../../img/DataTask.svg') var(--mask-icon-size);
  mask: url('../../../img/DataTask.svg') var(--mask-icon-size);
}

div[class*='bpmn-icon-confirmation-task'] {
  -webkit-mask: url('../../../img/ConfirmationTask.svg') var(--mask-icon-size);
  mask: url('../../../img/ConfirmationTask.svg') var(--mask-icon-size);
}

div[class*='bpmn-icon-feedback-task'] {
  -webkit-mask: url('../../../img/FeedbackTask.svg') var(--mask-icon-size);
  mask: url('../../../img/FeedbackTask.svg') var(--mask-icon-size);
}

div[class*='bpmn-icon-signing-task'] {
  -webkit-mask: url('../../../img/SignTask.svg') var(--mask-icon-size);
  mask: url('../../../img/SignTask.svg') var(--mask-icon-size);
}
