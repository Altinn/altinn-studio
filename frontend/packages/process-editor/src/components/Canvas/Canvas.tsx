import React, { useState } from 'react';

import 'bpmn-js/dist/assets/diagram-js.css';
import 'bpmn-js/dist/assets/bpmn-js.css';
import 'bpmn-js/dist/assets/bpmn-font/css/bpmn-embedded.css';

import classes from './Canvas.module.css';
import { useBpmnContext } from '../../contexts/BpmnContext';
import { shouldDisplayFeature } from 'app-shared/utils/featureToggleUtils';
import { supportsProcessEditor } from '../../utils/processEditorUtils';
import { BPMNViewer } from './BPMNViewer';
import { BPMNEditor } from './BPMNEditor';
import { CanvasActionMenu } from './CanvasActionMenu';

export type CanvasProps = {
  onSave: (bpmnXml: string) => void;
  appLibVersion: string;
};

/**
 * @component
 *  Displays the canvas area of the ProcessEditor
 *
 * @property {function}[onSave] - Function to be executed when saving the canvas
 * @property {string}[appLibVersion] - The app-lib version the user has
 *
 * @returns {ReactNode} - The rendered component
 */
export const Canvas = ({ onSave, appLibVersion }: CanvasProps): JSX.Element => {
  const { getUpdatedXml } = useBpmnContext();
  const [isEditorView, setIsEditorView] = useState(false);

  const toggleViewModus = (): void => {
    setIsEditorView((prevIsEditorView) => !prevIsEditorView);
  };

  const handleOnSave = async (): Promise<void> => {
    onSave(await getUpdatedXml());
  };

  const isEditAllowed: boolean = supportsProcessEditor(appLibVersion);

  return (
    <>
      {(isEditAllowed || shouldDisplayFeature('shouldOverrideAppLibCheck')) && (
        <CanvasActionMenu
          onSave={handleOnSave}
          toggleViewModus={toggleViewModus}
          isEditorView={isEditorView}
        />
      )}
      <div className={classes.wrapper}>
        {isEditorView ? <BPMNEditor /> : <BPMNViewer appLibVersion={appLibVersion} />}
      </div>
    </>
  );
};
