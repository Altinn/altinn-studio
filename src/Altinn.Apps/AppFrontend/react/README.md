## Running the tests

### Lint checks

1. Navigate to the folder `src/Altinn.Apps/AppFrontend/react`.
2. Execute `npm ci`. This step is only nescessary if you have not already done it, or if you change branches.
3. Execute `npm run install-deps`. This step is only nescessary if you have not already done it, or if you change branches.
4. Execute `npm run lint`.

### Unit tests

1. Navigate to the folder `src/Altinn.Apps/AppFrontend/react`.
2. Execute `npm ci`. This step is only nescessary if you have not already done it, or if you change branches.
3. Execute `npm run install-deps`. This step is only nescessary if you have not already done it, or if you change branches.
4. Execute `npm run test`.
