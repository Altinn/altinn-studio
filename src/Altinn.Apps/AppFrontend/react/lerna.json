{
  "command": {
    "bootstrap": {
      "hoist": true
    }
  },
  "packages": [
    "./*"
  ],
  "version": "independent"
}
