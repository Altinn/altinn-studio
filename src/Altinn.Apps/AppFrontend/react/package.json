{
	"name": "app-frontend",
	"private": true,
	"scripts": {
		"install-deps": "lerna bootstrap --hoist --ci",
		"no-ci-hoist": "lerna bootstrap --hoist --no-ci",
		"install-deps-with-lock": "lerna bootstrap",
		"test": "lerna exec --concurrency 1 -- npm test",
		"install-all": "lerna exec -- npm ci",
		"clean-lock": "lerna exec -- rm ./package-lock.json",
		"clean-node-modules": "rm -rf ./node_modules && lerna exec --concurrency 1 -- rm -rf ./node_modules",
		"update-packages": "lerna exec --concurrency 1 -- npm update --dev",
		"list-outdated": "lerna run list-outdated"
	},
	"devDependencies": {
		"@typescript-eslint/eslint-plugin": "~4.8.2",
		"@typescript-eslint/parser": "~4.8.2",
		"eslint": "~7.14.0",
		"eslint-config-airbnb": "~18.2.1",
		"eslint-plugin-import": "~2.22.1",
		"eslint-plugin-jsx-a11y": "~6.4.1",
		"eslint-plugin-react": "~7.21.5",
		"eslint-plugin-react-hooks": "^4.2.0",
		"ini": "1.3.5",
		"lerna": "3.22.1"
	}
}
