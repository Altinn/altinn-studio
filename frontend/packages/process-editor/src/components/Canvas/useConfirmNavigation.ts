import { useCallback, useEffect } from 'react';
import { useBeforeUnload, useBlocker } from 'react-router-dom';

export const useConfirmNavigation = (hasUnsavedChanges: boolean, confirmationMessage: string) => {
  useBeforeUnload(
    useCallback(
      (event) => {
        if (hasUnsavedChanges) {
          event.preventDefault();
          event.returnValue = confirmationMessage;
        }
      },
      [confirmationMessage, hasUnsavedChanges],
    ),
    { capture: true },
  );

  const blocker = useBlocker(({ currentLocation, nextLocation }) => {
    return hasUnsavedChanges && currentLocation.pathname !== nextLocation.pathname;
  });

  useEffect(() => {
    if (blocker.state === 'blocked') {
      if (window.confirm(confirmationMessage)) {
        blocker.proceed();
      } else {
        blocker.reset();
      }
    }
  }, [blocker, confirmationMessage]);
};
