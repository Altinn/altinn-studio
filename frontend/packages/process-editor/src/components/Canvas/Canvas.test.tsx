import React  from 'react';
import { Canvas } from './Canvas';
import { render , screen } from '@testing-library/react';
import BpmnJs from 'bpmn-js/dist/bpmn-navigated-viewer.production.min';

const renderNoDiagramError = jest.fn();
const renderNoProcessError = jest.fn(); 

const mockBpmn = {
  get: jest.fn().mockReturnValue({
    zoom: jest.fn(),
    stepZoom: jest.fn(),
    add: jest.fn(),
    onSave: jest.fn(),
  }),
  destroy: jest.fn(),
  on: jest.fn(),
  importXML: jest.fn(),
};    

jest.mock('bpmn-js/dist/bpmn-navigated-viewer.production.min')
  BpmnJs.mockImplementation(() => mockBpmn);

jest.mock('../../hooks/useBpmnViewer', () => ({
  useBpmnViewer: () => ({
    canvasRef: { current: null },
    renderNoDiagramError: false,
    renderNoProcessError: false,
  }),
}));

describe('Canvas', () => {
  it('should render Canvas', () => {
    const { container } = render(<Canvas onSave={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  it('should not render the alert when there are diagram and process', async () => {
    render(<Canvas onSave={() => {}} />);
    renderNoDiagramError.mockReturnValue(false);
    renderNoProcessError.mockReturnValue(false);
    expect(screen.queryByText('process_editor.not_found_diagram_heading')).not.toBeInTheDocument();
    expect(screen.queryByText('process_editor.not_found_diagram_error_message')).not.toBeInTheDocument();
  });

});