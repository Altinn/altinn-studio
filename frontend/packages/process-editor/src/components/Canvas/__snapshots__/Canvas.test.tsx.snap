// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas should render Alert when either renderNoDiagramError or renderNoProcessError is true 1`] = `
<div>
  <span>
    <button
      class="Button-module_button__2ZuB7 utility-module_focusable__1fBKr Button-module_medium__KTxdk Button-module_outline__F5jq- Button-module_primary__s1sM6"
      type="button"
    >
      <svg
        class="Button-module_icon__-43u5"
        fill="none"
        focusable="false"
        height="1em"
        role="img"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10.543 2.25a.75.75 0 0 0-.729.573l-.459 1.89c-.217.08-.43.167-.637.265l-1.675-1.02a.75.75 0 0 0-.92.11L4.066 6.121a.75.75 0 0 0-.11.92l1.02 1.676c-.097.207-.185.42-.264.637l-1.89.46a.75.75 0 0 0-.573.728v2.914a.75.75 0 0 0 .573.729l1.89.459c.079.217.167.43.265.637l-1.02 1.675a.75.75 0 0 0 .11.92l2.054 2.056a.75.75 0 0 0 .92.11l1.676-1.02c.207.097.42.185.637.264l.46 1.89a.75.75 0 0 0 .728.573h2.914a.75.75 0 0 0 .729-.573l.459-1.89c.217-.079.43-.167.637-.264l1.675 1.02a.75.75 0 0 0 .92-.11l2.056-2.055a.75.75 0 0 0 .11-.92l-1.02-1.676a7.69 7.69 0 0 0 .264-.637l1.89-.46a.75.75 0 0 0 .573-.728v-2.914a.75.75 0 0 0-.573-.729l-1.89-.459a7.693 7.693 0 0 0-.265-.637l1.02-1.675a.75.75 0 0 0-.11-.92l-2.054-2.056a.75.75 0 0 0-.92-.11l-1.676 1.02a7.706 7.706 0 0 0-.637-.264l-.46-1.89a.75.75 0 0 0-.728-.573h-2.914Zm.172 3.221.418-1.721h1.734l.418 1.721a.75.75 0 0 0 .513.541c.407.122.796.285 1.163.483a.75.75 0 0 0 .747-.02l1.526-.93 1.221 1.22-.93 1.527a.75.75 0 0 0-.02.747c.198.367.36.756.483 1.163a.75.75 0 0 0 .541.513l1.721.418v1.734l-1.72.418a.75.75 0 0 0-.542.513 6.207 6.207 0 0 1-.483 1.163.75.75 0 0 0 .02.747l.93 1.526-1.22 1.221-1.527-.93a.75.75 0 0 0-.747-.02 6.21 6.21 0 0 1-1.163.483.75.75 0 0 0-.513.541l-.418 1.721h-1.734l-.418-1.72a.75.75 0 0 0-.513-.542 6.21 6.21 0 0 1-1.163-.483.75.75 0 0 0-.747.02l-1.526.93-1.221-1.22.93-1.527a.75.75 0 0 0 .02-.747 6.206 6.206 0 0 1-.483-1.163.75.75 0 0 0-.541-.513l-1.721-.418v-1.734l1.721-.418a.75.75 0 0 0 .541-.513c.122-.407.285-.796.483-1.163a.75.75 0 0 0-.02-.747l-.93-1.526 1.22-1.221 1.527.93a.75.75 0 0 0 .747.02c.367-.198.756-.36 1.163-.483a.75.75 0 0 0 .513-.541ZM10.25 12a1.75 1.75 0 1 1 3.5 0 1.75 1.75 0 0 1-3.5 0ZM12 8.75a3.25 3.25 0 1 0 0 6.5 3.25 3.25 0 0 0 0-6.5Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
      [mockedText(process_editor.edit_mode)]
    </button>
    <div />
  </span>
  <div />
</div>
`;

exports[`Canvas should render Canvas 1`] = `
<div>
  <span>
    <button
      class="Button-module_button__2ZuB7 utility-module_focusable__1fBKr Button-module_medium__KTxdk Button-module_outline__F5jq- Button-module_primary__s1sM6"
      type="button"
    >
      <svg
        class="Button-module_icon__-43u5"
        fill="none"
        focusable="false"
        height="1em"
        role="img"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10.543 2.25a.75.75 0 0 0-.729.573l-.459 1.89c-.217.08-.43.167-.637.265l-1.675-1.02a.75.75 0 0 0-.92.11L4.066 6.121a.75.75 0 0 0-.11.92l1.02 1.676c-.097.207-.185.42-.264.637l-1.89.46a.75.75 0 0 0-.573.728v2.914a.75.75 0 0 0 .573.729l1.89.459c.079.217.167.43.265.637l-1.02 1.675a.75.75 0 0 0 .11.92l2.054 2.056a.75.75 0 0 0 .92.11l1.676-1.02c.207.097.42.185.637.264l.46 1.89a.75.75 0 0 0 .728.573h2.914a.75.75 0 0 0 .729-.573l.459-1.89c.217-.079.43-.167.637-.264l1.675 1.02a.75.75 0 0 0 .92-.11l2.056-2.055a.75.75 0 0 0 .11-.92l-1.02-1.676a7.69 7.69 0 0 0 .264-.637l1.89-.46a.75.75 0 0 0 .573-.728v-2.914a.75.75 0 0 0-.573-.729l-1.89-.459a7.693 7.693 0 0 0-.265-.637l1.02-1.675a.75.75 0 0 0-.11-.92l-2.054-2.056a.75.75 0 0 0-.92-.11l-1.676 1.02a7.706 7.706 0 0 0-.637-.264l-.46-1.89a.75.75 0 0 0-.728-.573h-2.914Zm.172 3.221.418-1.721h1.734l.418 1.721a.75.75 0 0 0 .513.541c.407.122.796.285 1.163.483a.75.75 0 0 0 .747-.02l1.526-.93 1.221 1.22-.93 1.527a.75.75 0 0 0-.02.747c.198.367.36.756.483 1.163a.75.75 0 0 0 .541.513l1.721.418v1.734l-1.72.418a.75.75 0 0 0-.542.513 6.207 6.207 0 0 1-.483 1.163.75.75 0 0 0 .02.747l.93 1.526-1.22 1.221-1.527-.93a.75.75 0 0 0-.747-.02 6.21 6.21 0 0 1-1.163.483.75.75 0 0 0-.513.541l-.418 1.721h-1.734l-.418-1.72a.75.75 0 0 0-.513-.542 6.21 6.21 0 0 1-1.163-.483.75.75 0 0 0-.747.02l-1.526.93-1.221-1.22.93-1.527a.75.75 0 0 0 .02-.747 6.206 6.206 0 0 1-.483-1.163.75.75 0 0 0-.541-.513l-1.721-.418v-1.734l1.721-.418a.75.75 0 0 0 .541-.513c.122-.407.285-.796.483-1.163a.75.75 0 0 0-.02-.747l-.93-1.526 1.22-1.221 1.527.93a.75.75 0 0 0 .747.02c.367-.198.756-.36 1.163-.483a.75.75 0 0 0 .513-.541ZM10.25 12a1.75 1.75 0 1 1 3.5 0 1.75 1.75 0 0 1-3.5 0ZM12 8.75a3.25 3.25 0 1 0 0 6.5 3.25 3.25 0 0 0 0-6.5Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
      [mockedText(process_editor.edit_mode)]
    </button>
    <div />
  </span>
  <div />
</div>
`;

exports[`Canvas should render Viewer when there is a diagram and process 1`] = `undefined`;
